File: reactive-logger-example-java/src/main/java/io/github/numichi/reactive/logger/example/java/ExampleComponent.java
package io.github.numichi.reactive.logger.example.java;

import io.github.numichi.reactive.logger.MDC;
import io.github.numichi.reactive.logger.reactor.MDCContext;
import io.github.numichi.reactive.logger.reactor.ReactiveLogger;
import org.slf4j.Logger;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Signal;
import reactor.util.annotation.NonNull;

import java.util.HashMap;
import java.util.Map;

public class ExampleComponent {
    
    private final Map<String, String> map = new HashMap<>();
    private final ReactiveLogger reactiveLogger;
    
    public ExampleComponent(Logger logger) {
        this.reactiveLogger = ReactiveLogger.getLogger(logger);
        map.put("foo", "bar");
        map.put(null, null);
    }
    
    public Mono<Map<String, String>> example1() {
        return Mono.just("example1")
            .flatMap(data -> reactiveLogger.info(data).thenReturn(data))
            .flatMap(data -> MDCContext.snapshot())
            .map(MDC::getData)
            .contextWrite(context -> MDCContext.merge(context, map));
    }
    
    public Mono<Void> example2() {
        return MDCContext.read()
            .doOnEach(reactiveLogger.logOnEach((Logger logger, Signal<MDC> signal) -> {
                String foo = getFooValue(signal);
                
                // Ofc, you can replace foo by be any message
                if (signal.isOnSubscribe()) logger.trace(foo);
                if (signal.isOnNext()) logger.debug(foo);
                if (signal.isOnComplete()) logger.info(foo);
                if (signal.isOnError()) logger.warn(foo);
            }))
            .contextWrite(context -> MDCContext.merge(context, map))
            .then();
    }
    
    public Mono<Void> example3() {
        return MDCContext.read()
            .doOnEach(signal -> reactiveLogger.logOnSignal(signal, (Logger logger) -> {
                String foo = getFooValue(signal);
                
                // Ofc, you can replace foo by be any message
                if (signal.isOnSubscribe()) logger.trace(foo);
                if (signal.isOnNext()) logger.debug(foo);
                if (signal.isOnComplete()) logger.info(foo);
                if (signal.isOnError()) logger.warn(foo);
            }))
            .contextWrite(context -> MDCContext.merge(context, map))
            .then();
    }
    
    public Mono<MDC> example4() {
        return MDCContext.read()
            .doOnEach(reactiveLogger.logOnEach((Logger logger, Signal<MDC> signal) -> {
                if (signal.isOnError()) {
                    logger.error(signal.getThrowable().getMessage());
                }
            }));
    }
    
    private String getFooValue(@NonNull Signal<MDC> signal)  {
        var mdc = signal.get();
        String foo = "null";
        
        if (mdc != null) {
            foo = mdc.getData().get("foo");
        }
        
        return foo;
    }
}

File: reactive-logger-example-kotlin/src/main/kotlin/io/github/numichi/reactive/logger/example/kotlin/ExampleComponent.kt
package io.github.numichi.reactive.logger.example.kotlin

import io.github.numichi.reactive.logger.MDC
import io.github.numichi.reactive.logger.coroutine.CoroutineLogger
import io.github.numichi.reactive.logger.coroutine.modifyMdc
import io.github.numichi.reactive.logger.coroutine.readMdc
import io.github.numichi.reactive.logger.coroutine.snapshotMdc
import io.github.numichi.reactive.logger.coroutine.withMDCContext
import io.github.numichi.reactive.logger.exceptions.ReadException
import io.github.numichi.reactive.logger.reactor.MDCContext
import kotlinx.coroutines.reactor.awaitSingleOrNull
import org.slf4j.Logger
import reactor.core.publisher.Signal

class ExampleComponent(logger: Logger) {

    private val map = mapOf("foo" to "bar")
    private val coroutineLogger = CoroutineLogger.getLogger(logger)

    suspend fun example1(): Map<String, String?> {
        return withMDCContext(MDC(map)) {
            coroutineLogger.info("example1")
            snapshotMdc().data
        }
    }

    suspend fun example2() {
        withMDCContext(MDC(map)) {
            val mdc = readMdc()
            coroutineLogger.debug(mdc.data["foo"])
        }

        // ========= OR =========

        MDCContext.read() // Mono<MDC>
            .doOnEach {
                coroutineLogger.logOnSignal(it) { logger ->
                    val foo = getFooValue(it)

                    // Ofc, you can replace foo by be any message
                    if (it.isOnSubscribe) logger.trace(foo)
                    if (it.isOnNext) logger.debug(foo)
                    if (it.isOnComplete) logger.info(foo)
                    if (it.isOnError) logger.warn(foo)
                }
            }
            .contextWrite { it.modifyMdc(map) }
            .awaitSingleOrNull()
    }

    suspend fun example3() {
        withMDCContext(MDC(map)) {
            val mdc = readMdc()
            coroutineLogger.debug(mdc.data["foo"])
        }

        // ========= OR =========

        MDCContext.read() // Mono<MDC>
            .doOnEach(coroutineLogger.logOnEach { logger, signal ->
                val foo = getFooValue(signal)

                // Ofc, you can replace foo by be any message
                if (signal.isOnSubscribe) logger.trace(foo)
                if (signal.isOnNext) logger.debug(foo)
                if (signal.isOnComplete) logger.info(foo)
                if (signal.isOnError) logger.warn(foo)
            })
            .contextWrite { it.modifyMdc(map) }
            .awaitSingleOrNull()
    }

    suspend fun example4(): MDC {
        return try {
            readMdc()
        } catch (e: ReadException) {
            coroutineLogger.error(e.message)
            throw e
        }
    }

    private fun getFooValue(signal: Signal<MDC>): String? {
        val mdc = signal.get()
        var foo: String? = "null"
        if (mdc != null) {
            foo = mdc.data["foo"]
        }
        return foo
    }
}